File: leetcode2/MinimizeDeviationinArray/a.cpp
/* Copyright (C) 2020 by iamslash */

#include <cstdio>
#include <vector>
#include <set>

using namespace std;

// nums: 1 2 3 4
//  ans:            3
//
//       2 2 4 6
//       2 2 3 4    2
//       2 2 2 3    1

// nums: 4 1 5 20 3    ans
//
//       1 3 4 5 20    
//       2 4 6 10 20   18
//       2 4 6 10 10   8
//       2 4 5  6 10   8
//       2 4 5  5  6   4
//       2 3 4  5  5   3

// 900ms 21.47% 121.7MB 34.53%
// priority queue
// O(NlgN) O(N)
class Solution {
 public:
  int minimumDeviation(vector<int>& nums) {
    set<int> numSet;
    for (int num : nums) {
      numSet.insert(num % 2 > 0 ? num * 2 : num);
    }
    int ans = *numSet.rbegin() - *numSet.begin();
    while (*numSet.rbegin() % 2 == 0) {
      numSet.insert(*numSet.rbegin() / 2);
      numSet.erase(*numSet.rbegin());
      ans = min(ans, *numSet.rbegin() - *numSet.begin());
    }
    return ans;
  }
};

File: leetcode2/MinimizeDeviationinArray/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2020 by iamslash

from typing import List
import heapq

# 900ms 87.33% 26.3MB 75.11%
# priority queue
# O(NlgN) O(N)
class Solution:
    def minimumDeviation(self, nums: List[int]) -> int:
        pq = []
        for a in nums:
            heapq.heappush(pq, -a * 2 if a % 2 > 0 else -a)
        ans = float('inf')
        minNum = -max(pq)
        while True:
            # print(pq)
            # print(minNum)          
            num = -heapq.heappop(pq)
            ans = min(ans, num - minNum)
            if num % 2 > 0:
                break
            num //= 2
            minNum = min(minNum, num)
            heapq.heappush(pq, -num)
        return ans
                
if __name__ == "__main__":
    sln = Solution()
    # 1
    print(sln.minimumDeviation([1,2,3,4]))
    # 3
    print(sln.minimumDeviation([4,1,5,20,3]))
    # 3
    print(sln.minimumDeviation([2,10,8]))
